compiler/src/lnntoamm/Expr.ts

```diff
@@ -565,80 +590,62 @@ class Call extends Expr {
     }
     // first reduction
     const argTys = args.map((arg) => arg.ty);
-    const selFns = FunctionType.matrixSelect(fns, argTys, metadata.scope);
-    fns = selFns.map((selFn) => selFn[0]);
-    const retTy = Type.oneOf(selFns.map(([_fn, ty]) => ty));
+    const retTy = Type.generate();
+    const [selFns, selPTys, selRetTys] = FunctionType.matrixSelect(
+      fns,
+      argTys,
+      retTy,
+      metadata.scope,
+    );
+    fns = selFns;
+    // console.log('initial select for', ast.t.trim(), '->', fns);
+    // stdout.write('-> args ');
+    // console.dir(args, { depth: 4 });
+    args.forEach((arg, ii) =>
+      arg.ty.constrain(Type.oneOf(selPTys[ii]), metadata.scope),
+    );
+    retTy.constrain(Type.oneOf(selRetTys), metadata.scope);
     // now, constrain all of the args to their possible types
     // makes it so that the type of the parameters in each position are in their own list
     // ie, given `do(int8, int16)` and `do(int8, int8)`, will result in this 2D array:
     // [ [int8, int8],
     //   [int16, int8] ]
     // for some reason TS thinks that `fns` is `Boxish` but *only* in the lambda here,
     // which is why I have to specify `fns: Fn[]`...
-    argTys.forEach((ty, ii) => {
-      const paramTys = (fns as Fn[]).map((fn) => fn.params[ii].ty);
-      // console.log('constraining', ty, 'to', paramTys);
-      ty.constrain(Type.oneOf(paramTys), metadata.scope);
-      // console.log('constrained:', ty);
-    });
+    // argTys.forEach((ty, ii) => {
+    //   const paramTys = (fns as Fn[]).map((fn) => fn.params[ii].ty);
+    //   // console.log('constraining', ty, 'to', paramTys);
+    //   ty.constrain(Type.oneOf(paramTys), metadata.scope);
+    //   // console.log('constrained:', ty);
+    // });
     if (closure !== null) {
       TODO('closures');
     }
     return [stmts, new Call(ast, fns, closure, args, metadata.scope, retTy)];
   }
 
-  private fnSelect(): [Fn[], Type[]] {
-    return FunctionType.matrixSelect(
+  private fnSelect(): [Fn[], Type[][], Type[]] {
+    const ret = FunctionType.matrixSelect(
       this.fns,
       this.args.map((a) => a.ty),
+      this.retTy,
       this.scope,
-    ).reduce(
-      ([fns, tys], [fn, ty]) => [
-        [...fns, fn],
-        [...tys, ty],
-      ],
-      [new Array<Fn>(), new Array<Type>()],
     );
+    // console.log('fnSelect for', this.ast.t.trim(), '->', ret[0]);
```

Comment:
```suggestion
```

Reply:
this is also useful, so i was planning on keeping it